File: backend/app/models/enums.py
"""Canonical enumerations.

Stored as TEXT + CHECK rather than as PostgreSQL ENUM types, because:

  * WatermelonDB has no enum type -- the value crosses the wire as a plain
    string regardless, so a native PG enum buys nothing at the boundary;
  * extending a native enum needs an ALTER TYPE migration, which is awkward to
    keep in lockstep with a mobile schema version bump.

`tests/test_schema_integrity.py` asserts that every CHECK constraint in the
live database matches the tuples here, so a value added on one side and
forgotten on the other fails CI rather than production.

Vietnamese labels live here too, and are served to the mobile app so the two
never disagree about what `land_prep` is called.
"""

from __future__ import annotations

from enum import Enum


class StrEnum(str, Enum):
    """`str`-valued enum. (`enum.StrEnum` proper is 3.11+.)"""

    def __str__(self) -> str:  # pragma: no cover - trivial
        return str(self.value)

    @classmethod
    def values(cls) -> tuple[str, ...]:
        return tuple(m.value for m in cls)


class WorkType(StrEnum):
    LAND_PREP = "land_prep"
    SOWING = "sowing"
    FERTILIZING = "fertilizing"
    SPRAYING = "spraying"
    WATERING = "watering"
    WEEDING = "weeding"
    HARVESTING = "harvesting"
    OTHER = "other"


class SupplyCategory(StrEnum):
    FERTILIZER = "fertilizer"
    PESTICIDE = "pesticide"
    SEED = "seed"
    FUEL = "fuel"
    TOOL = "tool"
    OTHER = "other"


class TxnType(StrEnum):
    IN = "in"
    OUT = "out"
    ADJUST = "adjust"


class ExpenseCategory(StrEnum):
    SUPPLY = "supply"
    LABOR = "labor"
    MACHINERY = "machinery"
    TRANSPORT = "transport"
    LAND_RENT = "land_rent"
    IRRIGATION = "irrigation"
    OTHER = "other"


class ExpenseSource(StrEnum):
    MANUAL = "manual"
    DIARY_AUTO = "diary_auto"


class SeasonStatus(StrEnum):
    PLANNING = "planning"
    ACTIVE = "active"
    HARVESTED = "harvested"
    CLOSED = "closed"


class SyncDirection(StrEnum):
    PULL = "pull"
    PUSH = "push"


class SyncStatus(StrEnum):
    OK = "ok"
    PARTIAL = "partial"
    ERROR = "error"


# ═══════════════════════════════════════════════════════════════════════════
#  Vietnamese UI labels
# ═══════════════════════════════════════════════════════════════════════════

WORK_TYPE_LABELS_VI: dict[str, str] = {
    WorkType.LAND_PREP: "Làm đất",
    WorkType.SOWING: "Gieo/Trồng",
    WorkType.FERTILIZING: "Bón phân",
    WorkType.SPRAYING: "Phun thuốc",
    WorkType.WATERING: "Tưới nước",
    WorkType.WEEDING: "Làm cỏ",
    WorkType.HARVESTING: "Thu hoạch",
    WorkType.OTHER: "Khác",
}

SUPPLY_CATEGORY_LABELS_VI: dict[str, str] = {
    SupplyCategory.FERTILIZER: "Phân bón",
    SupplyCategory.PESTICIDE: "Thuốc BVTV",
    SupplyCategory.SEED: "Giống",
    SupplyCategory.FUEL: "Nhiên liệu",
    SupplyCategory.TOOL: "Dụng cụ",
    SupplyCategory.OTHER: "Khác",
}

TXN_TYPE_LABELS_VI: dict[str, str] = {
    TxnType.IN: "Nhập kho",
    TxnType.OUT: "Xuất kho",
    TxnType.ADJUST: "Điều chỉnh",
}

EXPENSE_CATEGORY_LABELS_VI: dict[str, str] = {
    ExpenseCategory.SUPPLY: "Vật tư",
    ExpenseCategory.LABOR: "Nhân công",
    ExpenseCategory.MACHINERY: "Máy móc",
    ExpenseCategory.TRANSPORT: "Vận chuyển",
    ExpenseCategory.LAND_RENT: "Thuê đất",
    ExpenseCategory.IRRIGATION: "Thủy lợi",
    ExpenseCategory.OTHER: "Khác",
}

SEASON_STATUS_LABELS_VI: dict[str, str] = {
    SeasonStatus.PLANNING: "Chuẩn bị",
    SeasonStatus.ACTIVE: "Đang canh tác",
    SeasonStatus.HARVESTED: "Đã thu hoạch",
    SeasonStatus.CLOSED: "Đã kết thúc",
}


def sql_in_list(values: tuple[str, ...]) -> str:
    """Render a tuple as a SQL IN-list literal for a CHECK constraint."""
    return ", ".join(f"'{v}'" for v in values)

File: mobile/src/db/enums.ts
/**
 * Canonical enumerations — mirrors backend/app/models/enums.py.
 *
 * The values are the contract; the labels are what the farmer reads. Both
 * live here so the two sides cannot disagree about what `land_prep` is
 * called. A backend test asserts every CHECK constraint in PostgreSQL matches
 * the Python tuples, and the values below must match those in turn.
 */

export const WorkType = {
  LAND_PREP: 'land_prep',
  SOWING: 'sowing',
  FERTILIZING: 'fertilizing',
  SPRAYING: 'spraying',
  WATERING: 'watering',
  WEEDING: 'weeding',
  HARVESTING: 'harvesting',
  OTHER: 'other',
} as const;
export type WorkType = (typeof WorkType)[keyof typeof WorkType];

export const SupplyCategory = {
  FERTILIZER: 'fertilizer',
  PESTICIDE: 'pesticide',
  SEED: 'seed',
  FUEL: 'fuel',
  TOOL: 'tool',
  OTHER: 'other',
} as const;
export type SupplyCategory = (typeof SupplyCategory)[keyof typeof SupplyCategory];

export const TxnType = {
  IN: 'in',
  OUT: 'out',
  ADJUST: 'adjust',
} as const;
export type TxnType = (typeof TxnType)[keyof typeof TxnType];

export const ExpenseCategory = {
  SUPPLY: 'supply',
  LABOR: 'labor',
  MACHINERY: 'machinery',
  TRANSPORT: 'transport',
  LAND_RENT: 'land_rent',
  IRRIGATION: 'irrigation',
  OTHER: 'other',
} as const;
export type ExpenseCategory = (typeof ExpenseCategory)[keyof typeof ExpenseCategory];

export const ExpenseSource = {
  MANUAL: 'manual',
  DIARY_AUTO: 'diary_auto',
} as const;
export type ExpenseSource = (typeof ExpenseSource)[keyof typeof ExpenseSource];

export const SeasonStatus = {
  PLANNING: 'planning',
  ACTIVE: 'active',
  HARVESTED: 'harvested',
  CLOSED: 'closed',
} as const;
export type SeasonStatus = (typeof SeasonStatus)[keyof typeof SeasonStatus];

// ─── Nhãn tiếng Việt ────────────────────────────────────────────────────────

export const WORK_TYPE_LABELS: Record<WorkType, string> = {
  land_prep: 'Làm đất',
  sowing: 'Gieo/Trồng',
  fertilizing: 'Bón phân',
  spraying: 'Phun thuốc',
  watering: 'Tưới nước',
  weeding: 'Làm cỏ',
  harvesting: 'Thu hoạch',
  other: 'Khác',
};

export const SUPPLY_CATEGORY_LABELS: Record<SupplyCategory, string> = {
  fertilizer: 'Phân bón',
  pesticide: 'Thuốc BVTV',
  seed: 'Giống',
  fuel: 'Nhiên liệu',
  tool: 'Dụng cụ',
  other: 'Khác',
};

export const TXN_TYPE_LABELS: Record<TxnType, string> = {
  in: 'Nhập kho',
  out: 'Xuất kho',
  adjust: 'Điều chỉnh',
};

export const EXPENSE_CATEGORY_LABELS: Record<ExpenseCategory, string> = {
  supply: 'Vật tư',
  labor: 'Nhân công',
  machinery: 'Máy móc',
  transport: 'Vận chuyển',
  land_rent: 'Thuê đất',
  irrigation: 'Thủy lợi',
  other: 'Khác',
};

export const SEASON_STATUS_LABELS: Record<SeasonStatus, string> = {
  planning: 'Chuẩn bị',
  active: 'Đang canh tác',
  harvested: 'Đã thu hoạch',
  closed: 'Đã kết thúc',
};

export const AREA_UNITS = ['sao', 'ha', 'm2', 'công', 'mẫu'] as const;
export const SUPPLY_UNITS = ['kg', 'g', 'tấn', 'L', 'ml', 'bao', 'chai', 'gói', 'cái', 'bình'] as const;
export const REVENUE_UNITS = ['kg', 'tạ', 'tấn', 'bao', 'thùng', 'quả', 'bó'] as const;
export const WEATHER_VALUES = ['sunny', 'cloudy', 'rain', 'storm', 'windy'] as const;

export const WEATHER_LABELS: Record<string, string> = {
  sunny: 'Nắng',
  cloudy: 'Nhiều mây',
  rain: 'Mưa',
  storm: 'Giông bão',
  windy: 'Gió mạnh',
};
